libslide/src/partial_evaluator/flatten.rs

```diff
@@ -155,6 +172,318 @@ fn flatten_add_or_sub(o_lhs: &Rc<Expr>, o_rhs: &Rc<Expr>, is_subtract: bool) ->
+    let mut args_before_div = if is_div { 1 } else { 2 };
+
+    while let Some(arg) = args.pop_front() {
+        let div_side = args_before_div <= 0;
+        args_before_div -= 1;
+
+        let arg = unwrap_expr(arg);
+
+        match arg.as_ref() {
+            Expr::Const(konst) => {
+                if div_side {
+                    coeff /= konst;
+                } else {
+                    coeff *= konst;
+                }
+            }
+            Expr::BinaryExpr(BinaryExpr { op, lhs, rhs })
+                if op == &BinaryOperator::Mult || op == &BinaryOperator::Div =>
+            {
+                if div_side {
+                    if op == &BinaryOperator::Mult {
+                        // 1 / (2 * 3) -> 1 / 2 / 3; add both operands to the div side.
+                        args.push_back(lhs);
+                        args.push_back(rhs);
+                    } else {
+                        // 1 / (2 / 3) -> 3 / 2; here we multiply by the reciprocal, so 3 goes on
+                        // the mul side and 2 goes on the div side.
+                        args.push_front(rhs);
+                        args_before_div = 1;
+                        args.push_back(lhs);
+                    }
+                } else {
+                    // mul side
+                    if op == &BinaryOperator::Mult {
+                        // 1 * (2 * 3) -> 1 * 2 * 3
+                        args.push_front(lhs);
+                        args.push_front(rhs);
+                        args_before_div += 2;
+                    } else {
+                        // 1 * (2 / 3) -> 1 * 2 / 3; add 2 to the mul side and 3 to the div side.
+                        args.push_front(lhs);
+                        args_before_div += 1;
+                        args.push_back(rhs);
+                    }
+                }
+            }
+            _ => {
+                // Otherwise the arg is something we cannot further decompose in this context
+                // (e.g. a variable or an exponentiation), so add it as a term.
+                // TODO: see if we can handle other things more granularly
+                let entry = terms.entry(arg).or_insert(0.);
+                if div_side {
+                    *entry -= 1.;
+                } else {
+                    *entry += 1.;
+                }
+            }
+        }
+    }
+
+    let mut new_args: Vec<Rc<Expr>> = Vec::with_capacity(1 + terms.len());
+    if (coeff - 1.).abs() >= std::f64::EPSILON {
+        // coeff != 1
+        new_args.push(Rc::from(Expr::Const(coeff)));
+    }
+    for (term, coeff) in terms {
+        if coeff == 0. {
+            // The happiest path :)
+            continue;
+        } else if (coeff - 1.).abs() < std::f64::EPSILON {
+            // 1 * x
+            new_args.push(Rc::clone(term));
+        } else if (coeff - -1.).abs() < std::f64::EPSILON {
+            // -1 * x ~ 1/x
```

Comment:
can you explain the double negative here.

Reply:
We are checking if coeff == -1 relative to some epsilon, since we are dealing with machine precision. added a comment

Reply:
can't you just write + 1 though

Reply:
@ayazhafiz

Reply:
yeah but imo it's more clear this way, usually you subtract to compare two items and `-1` is explicit here